$ \Huge{solution \ \ of \ \ 2.4}$

let $x_1,x_2,x_3,x_4,x_5 $ be the number of drone of model SHREY,DHARAN,SATVA,ADHYA,PRAVAT respectively

we have to maximize possible carrying capacity so our objective function is


max $ 11x_1+13x_2 +9x_3+ 15x_4+12x_5$

subject to constraint

 $ 6.5x_1+8x_2 +10.5x_3+ 11.75x_4+14.5x_5 \le 78$


and    $ 5.75x_1+4.5x_2 +3.75x_3+ 4.85x_4+3.45x_5 \le 84$

since $x_1,x_2,x_3,x_4,x_5 $ is number of drone so, it can't be float it must be Non Negative Integer

In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import * 

In [ ]:
import numpy as np

In [ ]:
import pandas as pd

In [ ]:
data_csvfile = pd.read_csv('lab4_ex3.csv')

In [ ]:
data_csvfile.columns

Index(['Drone Model', 'Carrying Capacity (kgs)', 'Cost (in Crores of rupees)',
       'Parking space (sq. m.)'],
      dtype='object')

In [ ]:
data_csvfile['Drone Model']

0     SHREY
1    DHARAN
2     SATVA
3     ADHYA
4    PRAVAT
Name: Drone Model, dtype: object

In [ ]:
data_csvfile['Carrying Capacity (kgs)']

0    11
1    13
2     9
3    15
4    12
Name: Carrying Capacity (kgs), dtype: int64

In [ ]:
data_csvfile['Cost (in Crores of rupees)']

0     6.50
1     8.00
2    10.50
3    11.75
4    14.50
Name: Cost (in Crores of rupees), dtype: float64

In [ ]:
data_csvfile['Parking space (sq. m.)']

0    5.75
1    4.50
2    3.75
3    4.85
4    3.45
Name: Parking space (sq. m.), dtype: float64

In [ ]:
# create a model
model_lab4_ex2 = ConcreteModel()

In [ ]:
N = len(data_csvfile.index) 
#we declare a Python variable N denoting the number of decision variables  

In [ ]:
M =2 #we declare a Python variable M denoting the number of constraints

In [ ]:
col_indices = range(N)

In [ ]:
#declare the decision variables in the model
model_lab4_ex2.x = Var(col_indices)

In [ ]:
model_lab4_ex2.x.domain=NonNegativeIntegers

In [ ]:
model_lab4_ex2.objective = Objective(expr=summation(data_csvfile['Carrying Capacity (kgs)'], model_lab4_ex2.x),sense=maximize)

In [ ]:
model_lab4_ex2.constraints = ConstraintList()

In [ ]:
#adding coefficient from csv file
model_lab4_ex2.constraints.add(summation(data_csvfile['Cost (in Crores of rupees)'], model_lab4_ex2.x) <= 78)


In [ ]:

model_lab4_ex2.constraints.add(summation(data_csvfile['Parking space (sq. m.)'], model_lab4_ex2.x) <= 84 )


In [ ]:
#Since the variables have the same upper bound and lower bound, we can simply use scalars to store them
lower_bound = 0
upper_bound = np.inf

In [ ]:
#add the upper bound and lower bounds for the variables
for j in col_indices:
  model_lab4_ex2.x[j].setlb(lower_bound)
  model_lab4_ex2.x[j].setub(upper_bound)

In [ ]:
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeIntegers
          1 :     0 :  None :  None : False :  True : NonNegativeIntegers
          2 :     0 :  None :  None : False :  True : NonNegativeIntegers
          3 :     0 :  None :  None : False :  True : NonNegativeIntegers
          4 :     0 :  None :  None : False :  True : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3]

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('number  of drone model',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()


Objective =  132.0

Decision Variables
number  of drone model SHREY 12.0
number  of drone model DHARAN 0.0
number  of drone model SATVA 0.0
number  of drone model ADHYA 0.0
number  of drone model PRAVAT 0.0

Constraints
constraints : Size=2
    Key : Lower : Body : Upper
      1 :  None : 78.0 :  78.0
      2 :  None : 69.0 :  84.0


$ \huge{solution  \ \ of \ \ 2.6}$


optimal objective function value= 132

number  of drone model SHREY = 12.0

number  of drone model DHARAN = 0.0

number  of drone model SATVA = 0.0

number  of drone model ADHYA  = 0.0

number  of drone model PRAVAT  = 0.0

In [ ]:
model_lab4_ex2.x.domain=Reals

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('number of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()


Objective =  132.0

Decision Variables
number of drone model  SHREY 12.0
number of drone model  DHARAN 0.0
number of drone model  SATVA 0.0
number of drone model  ADHYA 0.0
number of drone model  PRAVAT 0.0

Constraints
constraints : Size=2
    Key : Lower : Body : Upper
      1 :  None : 78.0 :  78.0
      2 :  None : 69.0 :  84.0


$ \huge{solution  \ \ of \ \  2.7}$


optimal objective function value= 132

number  of drone model SHREY = 12.0

number  of drone model DHARAN = 0.0

number  of drone model SATVA = 0.0

number  of drone model ADHYA  = 0.0

number  of drone model PRAVAT  = 0.0

here solution is same for both LP and MILP but it is not always happend

$ \huge{solution \ \  of \ \ 2.8}$ 

yes here solution is same for both LP and MILP since it satisfy the theorem

If the optimal solution of the LP relaxation is feasible for the original MILP, then it is
optimal also for the original problem



$\huge{2.9}$

In [ ]:
model_lab4_ex2.constraints.add(summation(data_csvfile['Cost (in Crores of rupees)'],model_lab4_ex2.x) <= 76)

In [ ]:
model_lab4_ex2.constraints.add(summation(data_csvfile['Parking space (sq. m.)'],model_lab4_ex2.x) <= 87)

In [ ]:
model_lab4_ex2.x.domain=Reals

In [ ]:
model_lab4_ex2.constraints[1].deactivate()

In [ ]:
model_lab4_ex2.constraints[2].deactivate()

In [ ]:
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  12.0 :  None : False : False :  Reals
          1 :     0 :   0.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declarations
    constraints : Siz

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('number  of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()


Objective =  128.615388

Decision Variables
number  of drone model  SHREY 11.692308
number  of drone model  DHARAN 0.0
number  of drone model  SATVA 0.0
number  of drone model  ADHYA 0.0
number  of drone model  PRAVAT 0.0

Constraints
constraints : Size=4
    Key : Lower : Body              : Upper
      3 :  None : 76.00000200000001 :  76.0
      4 :  None :         67.230771 :  87.0


In [ ]:
model_lab4_ex2.x.domain=NonNegativeIntegers

In [ ]:
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.692308 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('number of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()


Objective =  127.0

Decision Variables
number of drone model  SHREY 8.0
number of drone model  DHARAN 3.0
number of drone model  SATVA 0.0
number of drone model  ADHYA 0.0
number of drone model  PRAVAT 0.0

Constraints
constraints : Size=4
    Key : Lower : Body : Upper
      3 :  None : 76.0 :  76.0
      4 :  None : 59.5 :  87.0


$\huge{solution \ \ of \ \ 2.9}$

$ \Huge{solution \ \ of \ \ LP \ \ Model}$


Objective =  128.615388

Decision Variables

number  of drone model  SHREY 11.692308

number  of drone model  DHARAN 0.0

number  of drone model  SATVA 0.0

number  of drone model  ADHYA 0.0

number  of drone model  PRAVAT 0.0

$\Huge{solution \ \ of \ \  MILP \ \ Model}$

Objective =  127.0

Decision Variables

number of drone model  SHREY 8.0

number of drone model  DHARAN 3.0

number of drone model  SATVA 0.0

number of drone model  ADHYA 0.0

number of drone model  PRAVAT 0.0
  

$\textbf{here we get different solution for MILP and LP}$

because here LP solution is infeasible for our problem as number of drone model can't be float
so, we get different solution for MILP



$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le74 \ \ and \ \  parking \ space \le88 }$ 

$\huge{by \ \ LP }$

In [ ]:
model_lab4_ex2.constraints.add(summation(data_csvfile['Cost (in Crores of rupees)'],model_lab4_ex2.x) <= 74)
model_lab4_ex2.constraints.add(summation(data_csvfile['Parking space (sq. m.)'],model_lab4_ex2.x) <= 88)
model_lab4_ex2.constraints[3].deactivate()
model_lab4_ex2.constraints[4].deactivate()
model_lab4_ex2.x.domain=Reals
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   8.0 :  None : False : False :  Reals
          1 :     0 :   3.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declarations
    constraints

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('no of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()


Solver status: ok
Solver termination condition: optimal

Objective =  125.230765

Decision Variables
no of drone model  SHREY 11.384615
no of drone model  DHARAN 0.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=6
    Key : Lower : Body        : Upper
      5 :  None :  73.9999975 :  74.0
      6 :  None : 65.46153625 :  88.0


$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le74 \ \ and \ \  parking \ space \le88 }$ 

$\huge{by \ \ MILP }$

In [ ]:
model_lab4_ex2.x.domain=NonNegativeIntegers
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.384615 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('no of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()

Solver status: ok
Solver termination condition: optimal

Objective =  123.0

Decision Variables
no of drone model  SHREY 10.0
no of drone model  DHARAN 1.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=6
    Key : Lower : Body : Upper
      5 :  None : 73.0 :  74.0
      6 :  None : 62.0 :  88.0


$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le74 \ \ and \ \  parking \ space \le88 }$ 

$\huge{by \ \ LP }$


Objective =  125.230765

Decision Variables

no of drone model  SHREY 11.384615

no of drone model  DHARAN 0.0

no of drone model  SATVA 0.0

no of drone model  ADHYA 0.0

no of drone model  PRAVAT 0.0

$\huge{by  \ \ MILP}$

Objective =  123.0

Decision Variables

no of drone model  SHREY 10.0

no of drone model  DHARAN 1.0

no of drone model  SATVA 0.0

no of drone model  ADHYA 0.0

no of drone model  PRAVAT 0.0

$\textbf{here we get different solution for MILP and LP}$

because here LP solution is infeasible for our problem as number of drone model can't be float
so, we get different solution for MILP


$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le72 \ \ and \ \  parking \ space \le89 }$ 

$\huge{by \ \ LP }$

In [ ]:
model_lab4_ex2.constraints.add(summation(data_csvfile['Cost (in Crores of rupees)'],model_lab4_ex2.x) <= 72)
model_lab4_ex2.constraints.add(summation(data_csvfile['Parking space (sq. m.)'],model_lab4_ex2.x) <= 89)
model_lab4_ex2.constraints[5].deactivate()
model_lab4_ex2.constraints[6].deactivate()
model_lab4_ex2.x.domain=Reals
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    8 : {1, 2, 3, 4, 5, 6, 7, 8}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  10.0 :  None : False : False :  Reals
          1 :     0 :   1.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declarations
    const

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('no of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()

Solver status: ok
Solver termination condition: optimal

Objective =  121.84615300000002

Decision Variables
no of drone model  SHREY 11.076923
no of drone model  DHARAN 0.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=8
    Key : Lower : Body               : Upper
      7 :  None :         71.9999995 :  72.0
      8 :  None : 63.692307250000006 :  89.0


$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le72 \ \ and \ \  parking \ space \le89 }$ 

$\huge{by \ \ MILP }$

In [ ]:
model_lab4_ex2.x.domain=NonNegativeIntegers
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    8 : {1, 2, 3, 4, 5, 6, 7, 8}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.076923 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maxim

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('no of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()

Solver status: ok
Solver termination condition: optimal

Objective =  121.0

Decision Variables
no of drone model  SHREY 11.0
no of drone model  DHARAN 0.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=8
    Key : Lower : Body  : Upper
      7 :  None :  71.5 :  72.0
      8 :  None : 63.25 :  89.0


$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le72 \ \ and \ \  parking \ space \le89 }$ 

$\huge{by \ \ LP }$

Objective =  121.84615300000002

Decision Variables

no of drone model  SHREY 11.076923

no of drone model  DHARAN 0.0

no of drone model  SATVA 0.0

no of drone model  ADHYA 0.0

no of drone model  PRAVAT 0.0

$\huge{by \ \ MILP}$

Objective =  121.0

Decision Variables

no of drone model  SHREY 11.0

no of drone model  DHARAN 0.0

no of drone model  SATVA 0.0

no of drone model  ADHYA 0.0

no of drone model  PRAVAT 0.0

$\textbf{here we get different solution for MILP and LP}$

because here LP solution is infeasible for our problem as number of drone model can't be float
so, we get different solution for MILP

$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le70 \ \ and \ \  parking \ space \le90 }$ 

$\huge{by \ \ LP }$

In [ ]:
model_lab4_ex2.constraints.add(summation(data_csvfile['Cost (in Crores of rupees)'],model_lab4_ex2.x) <= 70)
model_lab4_ex2.constraints.add(summation(data_csvfile['Parking space (sq. m.)'],model_lab4_ex2.x) <= 90)
model_lab4_ex2.constraints[7].deactivate()
model_lab4_ex2.constraints[8].deactivate()
model_lab4_ex2.x.domain=Reals
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  11.0 :  None : False : False :  Reals
          1 :     0 :   0.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declarations
  

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('no of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()

Solver status: ok
Solver termination condition: optimal

Objective =  118.461541

Decision Variables
no of drone model  SHREY 10.769231
no of drone model  DHARAN 0.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=10
    Key : Lower : Body               : Upper
      9 :  None :         70.0000015 :  70.0
     10 :  None : 61.923078249999996 :  90.0


$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le70 \ \ and \ \  parking \ space \le90 }$ 

$\huge{by \ \ MILP }$

In [ ]:
model_lab4_ex2.x.domain=NonNegativeIntegers
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 10.769231 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True 

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('no of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()

Solver status: ok
Solver termination condition: optimal

Objective =  116.0

Decision Variables
no of drone model  SHREY 7.0
no of drone model  DHARAN 3.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=10
    Key : Lower : Body  : Upper
      9 :  None :  69.5 :  70.0
     10 :  None : 53.75 :  90.0


$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le70 \ \ and \ \  parking \ space \le90 }$ 

$\huge{by \ \ LP }$

Objective =  118.461541


Decision Variables

no of drone model  SHREY 10.769231

no of drone model  DHARAN 0.0

no of drone model  SATVA 0.0

no of drone model  ADHYA 0.0

no of drone model  PRAVAT 0.0

$\Huge{by \ \ MILP }$

Objective =  116.0

Decision Variables

no of drone model  SHREY 7.0

no of drone model  DHARAN 3.0

no of drone model  SATVA 0.0

no of drone model  ADHYA 0.0

no of drone model  PRAVAT 0.0

$\textbf{here we get different solution for MILP and LP}$

because here LP solution is infeasible for our problem as number of drone model can't be float
so, we get different solution for MILP

$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le68 \ \ and \ \  parking \ space \le91 }$ 

$\huge{by \ \ LP }$

In [ ]:
model_lab4_ex2.constraints.add(summation(data_csvfile['Cost (in Crores of rupees)'],model_lab4_ex2.x) <= 68)
model_lab4_ex2.constraints.add(summation(data_csvfile['Parking space (sq. m.)'],model_lab4_ex2.x) <= 91)
model_lab4_ex2.constraints[9].deactivate()
model_lab4_ex2.constraints[10].deactivate()
model_lab4_ex2.x.domain=Reals
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   7.0 :  None : False : False :  Reals
          1 :     0 :   3.0 :  None : False : False :  Reals
          2 :     0 :   0.0 :  None : False : False :  Reals
          3 :     0 :   0.0 :  None : False : False :  Reals
          4 :     0 :   0.0 :  None : False : False :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 Constraint Declara

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('no of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()

Solver status: ok
Solver termination condition: optimal

Objective =  115.07691799999999

Decision Variables
no of drone model  SHREY 10.461538
no of drone model  DHARAN 0.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=12
    Key : Lower : Body               : Upper
     11 :  None :          67.999997 :  68.0
     12 :  None : 60.153843499999994 :  91.0


$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le68 \ \ and \ \  parking \ space \le91 }$ 

$\huge{by \ \ MILP }$

In [ ]:
model_lab4_ex2.x.domain=NonNegativeIntegers
model_lab4_ex2.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 10.461538 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :

In [ ]:
result = opt_cbc.solve(model_lab4_ex2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('\nObjective = ', model_lab4_ex2.objective())

print('\nDecision Variables')
for i in col_indices:
  print('no of drone model ',data_csvfile['Drone Model'][i], model_lab4_ex2.x[i].value)

print('\nConstraints')
model_lab4_ex2.constraints.display()

Solver status: ok
Solver termination condition: optimal

Objective =  114.0

Decision Variables
no of drone model  SHREY 8.0
no of drone model  DHARAN 2.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=12
    Key : Lower : Body : Upper
     11 :  None : 68.0 :  68.0
     12 :  None : 55.0 :  91.0


$\Huge{solution \ \ of \ \ 2.10 \ \ when } \\   \Huge{cost \le68 \ \ and \ \  parking \ space \le91 }$ 

$\huge{by \ \ LP }$

Objective =  115.07691799999999

Decision Variables

no of drone model  SHREY 10.461538

no of drone model  DHARAN 0.0

no of drone model  SATVA 0.0

no of drone model  ADHYA 0.0

no of drone model  PRAVAT 0.0

$\huge{by \ \ MILP}$

Objective =  114.0

Decision Variables

no of drone model  SHREY 8.0

no of drone model  DHARAN 2.0

no of drone model  SATVA 0.0

no of drone model  ADHYA 0.0

no of drone model  PRAVAT 0.0

$\textbf{here we get different solution for MILP and LP}$

because here LP solution is infeasible for our problem as number of drone model can't be float
so, we get different solution for MILP

$\huge{ we \ \ can \ \ solve \ \ 2.9 \ \ and  \ \ 2.10 \ \ by \ \ using \ \ for \ \  loop \ \  as \ \  follows}$   

In [ ]:
a=np.arange(5)
b=np.array([Reals,NonNegativeIntegers])

In [ ]:
cost=np.array([76,74,72,70,68])
parking=np.array([87,88,89,90,91])

In [ ]:
for j in range(2):
  for i in range(5):
    model_lab4_ex2.constraints.add(summation(data_csvfile['Cost (in Crores of rupees)'],model_lab4_ex2.x) <= cost[i])
    model_lab4_ex2.constraints.add(summation(data_csvfile['Parking space (sq. m.)'],model_lab4_ex2.x) <= parking[i])
    model_lab4_ex2.constraints[2*i+10*j+11].deactivate()
    model_lab4_ex2.constraints[2*i+10*j+12].deactivate()
    model_lab4_ex2.x.domain=b[j]
    result = opt_cbc.solve(model_lab4_ex2)
    print('\n\n\n\n\n\n\n\n\nsolution of model',i+1+5*j)
    print('Solver status:', result.solver.status)
    print('Solver termination condition:',result.solver.termination_condition)
    print('\nObjective = ', model_lab4_ex2.objective())
    for k in col_indices:
      print('no of drone model ',data_csvfile['Drone Model'][k], model_lab4_ex2.x[k].value)
    print('\nConstraints')
    model_lab4_ex2.constraints.display()
    











solution of model 1
Solver status: ok
Solver termination condition: optimal

Objective =  128.615388
no of drone model  SHREY 11.692308
no of drone model  DHARAN 0.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=14
    Key : Lower : Body              : Upper
     13 :  None : 76.00000200000001 :  76.0
     14 :  None :         67.230771 :  87.0









solution of model 2
Solver status: ok
Solver termination condition: optimal

Objective =  125.230765
no of drone model  SHREY 11.384615
no of drone model  DHARAN 0.0
no of drone model  SATVA 0.0
no of drone model  ADHYA 0.0
no of drone model  PRAVAT 0.0

Constraints
constraints : Size=16
    Key : Lower : Body        : Upper
     15 :  None :  73.9999975 :  74.0
     16 :  None : 65.46153625 :  88.0









solution of model 3
Solver status: ok
Solver termination condition: optimal

Objective =  121.84615300000002
no of drone model  SHREY 11.076923
no of

$\Huge{ 2.10}$

$\Huge{pattern}$

optimal solution of MILP IS always less than optimal solution of LP

$\huge{observation}$

If the optimal solution of the LP relaxation is feasible for the original MILP, then it is
optimal also for the original problem

If the optimal solution of the LP relaxation is not feasible for the original MILP, then it is not
optimal for the original problem 
in this case optimal solution of MILP IS always less than optimal solution of LP  